components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+            List<String> permissions, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        RoleBasicInfo role = roleDAO.addRole(roleName, userList, groupList, tenantDomain);
+        try {
+            getUserAdminProxy().setRoleUIPermission(roleName, permissions.toArray(new String[0]));
+        } catch (UserAdminException e) {
+            throw new IdentityRoleManagementServerException(UNEXPECTED_SERVER_ERROR.getCode(),
+                    "An error occurred when setting permissions for the role: " + roleName, e);
+        }
+        return role;
+    }
+
+    @Override
+    public List<RoleBasicInfo> getRoles(int limit, int offset, String sortBy, String sortOrder, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.getRoles(limit, offset, sortBy, sortOrder, tenantDomain);
+    }
+
+    @Override
+    public List<RoleBasicInfo> getRoles(String filter, int limit, int offset, String sortBy, String sortOrder,
+            String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.getRoles(filter, limit, offset, sortBy, sortOrder, tenantDomain);
+    }
+
+    @Override
+    public Role getRole(String roleID, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        Role role = roleDAO.getRole(roleID, tenantDomain);
+        role.setPermissions(getPermissionListOfRole(roleID, tenantDomain));
+        return role;
+    }
+
+    @Override
+    public RoleBasicInfo updateRoleName(String roleID, String newRoleName, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.updateRoleName(roleID, newRoleName, tenantDomain);
+    }
+
+    @Override
+    public void deleteRole(String roleID, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        roleDAO.deleteRole(roleID, tenantDomain);
+    }
+
+    @Override
+    public List<UserBasicInfo> getUserListOfRole(String roleID, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.getUserListOfRole(roleID, tenantDomain);
+    }
+
+    @Override
+    public RoleBasicInfo updateUserListOfRole(String roleID, List<String> newUserList, List<String> deletedUserList,
+            String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        roleDAO.addUsersToRole(roleID, newUserList, tenantDomain);
+        return roleDAO.removeUsersFromRole(roleID, deletedUserList, tenantDomain);
```

Comment:
Should we do two operations here. Isn't it possible to handle the set to be updated at dao layer and optimize the call from there

Reply:
or optimize retrieving role name having a common logic

Reply:
Or optimize avoiding duplicated calls

Reply:
We will have to do INSERT and DELETE operations to add and remove users. So don't we need two calls for that?

Reply:
agree on that but shouldn't that be atomic. Not sure how the implementation should be. Better to reach for a db expert. Otherwise won't it have race conditions?

Reply:
Added a single method in the DAO layer to handle the `updateUserListOfRole` as discussed.